src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -47,6 +49,8 @@ internal class ProjectHotReloadSessionManager : IProjectHotReloadSessionManager
 
         public async Task ActivateSessionAsync(int processId, bool runningUnderDebugger)
         {
+            using AsyncSemaphore.Releaser semaphoreReleaser = await _semaphore.EnterAsync();
```

Comment:
This is cool! I hadn't seen something like this before where you use the lifetime of the variable (IDisposable) to be the unlocking mechanism. Pretty nifty.

Reply:
Yeah it's nice. Beyond the convenience, it also ensures that any exception thrown within the scope will release the semaphore, so that it doesn't remain locked indefinitely after an error occurs.